proposals/2024/05-10_OpenID-Connect.md

```diff
@@ -0,0 +1,482 @@
+This increases productivity, strengthens security, and eliminates outages caused by expired or incorrect credentials.
+
+2. If this is a redesign or refactor, what issues exist in the current implementation?
+
+N/A
+
+3. Are there any other workarounds, and if so, what are the drawbacks?
+
+This is a new feature. Users would still be able to use the existing password management features.
+
+4. Are there any related issues? Please provide them below if any exist.
+
+
+## Design
+
+**Example Pipeline**
+
+High level, this would be the necessary first step to receive an ID token for any Vela build.
+
+```yaml
+version: "1"
+
+steps:
+  - name: request token
+    image: alpine:latest
+    id_request: write  # becomes a claim in token `request`
+    commands:
+      - apk add curl
+      - 'curl -H "Authorization: Bearer $VELA_ID_TOKEN_REQUEST_TOKEN" $VELA_ID_TOKEN_REQUEST_URL'
+```
+
+
+**New Token Types**
+
+### ID_TOKEN_REQUEST_TOKEN
+
+Token is signed by _VELA_SERVER_PRIVATE_KEY_ using HS256 algorithm. It has the following claims:
+
+```json
+{
+  "build_id": 1,
+  "build_number": 1,
+  "build_sender": "EastonCrupper",
+  "repo": "EastonCrupper/myvela",
+  "token_type": "IDRequest",
+  "image": "alpine:latest",
+  "request": "write",
+  "sub": "repo:<repo_full_name>:ref:<build_ref>:<build_event>",
+  "exp": 1714586259,
+  "iat": 1714584159
+}
+```
+
+The token is injected into any step that requests it: `id_request: <anything>`. The worker will use the executor build token to retrieve the `ID_TOKEN_REQUEST_TOKEN`:
+```go
+// WORKER CODE — PlanStep
+	var requestToken string
+
+	if len(ctn.IDRequest) > 0 {
+		opts := vela.RequestTokenOptions{
+			Image:   ctn.Image,
+			Request: ctn.IDRequest,
+		}
+
+		tkn, _, err := c.Vela.Build.GetIDRequestToken(c.build.GetRepo().GetOrg(), c.build.GetRepo().GetName(), c.build.GetNumber(), opts)
+		if err != nil {
+			return err
+		}
+
+		requestToken = tkn.GetToken()
+	}
+
+	// update the step container environment
+	//
+	// https://pkg.go.dev/github.com/go-vela/worker/internal/step#Environment
+	err = step.Environment(ctn, c.build, _step, c.Version, requestToken)
+	if err != nil {
+		return err
+	}
+```
+
+The token will then be minted with the image and `id_request` value as claims. This presents further opportunity for security checks from services, as these claims are translated to the ID token.
+
+Retrieving the token involves hitting the endpoint $VELA_ID_TOKEN_REQUEST_URL (/api/v1/repos/ORG/REPO/builds/BUILD/id_token). This environment variable is injected at compile time just like other platform level env variables.
+
+Upon hitting that endpoint, the server will validate a few things:
+
+- The token is the expected token type
+- The build is the expected build (build_id)
+- The token is not expired
+- The token has indeed been signed by the Vela server
+- The associated build is actively running (leaked request tokens are not vulnerable post build).
```

Comment:
does this run the risk that an external service _using_ the token might not be done with it yet? maybe we give it a little bit of wiggle room?

im picturing a scenario where Vela hands off the workload to a different service and that service might take some time to provision resources or something.

or... are we expecting id_tokens to _always_ be generated within the build context and then the resulting token can be handed off to the external service?

Reply:
The ID token request token is bound by that rule yes because the request for an ID token should always occur during a running build. If there is a pass off that occurs, the new workload should be using the ID token not the ID request token